### PR TITLE
Fix #609 Restarts with Full/Empty GPUs

### LINKED ISSUE #609: Restart Simulations: GPUs with/without Particles Hang
### Update of the issue

Restarting simulations where some GPUs have particles and other have no particles are hanging.

**Verified with examples:**
- Bunch, no-particle-GPUs at "vacuum" GPUs around the bunch
- LWFA, no-particle-GPUs on the "first GPUs before the profile starts"

**Does the problem affect the `master` branch**?

- [x] ~~likely, must check~~ yep

---

### First found in the Bunch example

With the current `dev` the restart of the `examples/Bunch` does not work.

Running the simulation and dumping the restart files appear to work fine. However restarting it leads to a hanging simulation.
The picongpu simulation hangs right after printing:
```bash
...
[1,0]<stdout>:PIConGPUVerbose PHYSICS(1) | UNIT_EFIELD 2.6633e+13
[1,0]<stdout>:PIConGPUVerbose PHYSICS(1) | UNIT_BFIELD 88838
[1,0]<stdout>:PIConGPUVerbose PHYSICS(1) | UNIT_ENERGY 1.39727e-10
```
`[1,0]<stdout>:initialization time: ...` is never reached. 

In order to test this, I used the default `Bunch` example and added `--checkpoints 3000`.
I also removed all plugins and disabled radiation to strip the simulation of as much unneeded kernels as possible.

With `SPLASH_VERBOSE=100` enabled (thanks @f-schmitt-zih, that's an ubercool feature) the following output is produced last after restart:
```bash
...
[1,3]<stderr>:[SPLASH_LOG:3] DCDataSet::read (particles_info)
[1,3]<stderr>:[SPLASH_LOG:3] 
[1,3]<stderr>:
[1,3]<stderr>: ndims         = 1
[1,3]<stderr>: logical_size  = (32,1,1)
[1,3]<stderr>: physical_size = (32,1,1)
[1,3]<stderr>: dstBuffer     = (32,1,1)
[1,3]<stderr>: dstOffset     = (0,0,0)
[1,3]<stderr>: srcSize       = (32,1,1)
[1,3]<stderr>: srcOffset     = (0,0,0)
[1,3]<stderr>:
[1,0]<stderr>:[SPLASH_LOG:0]  returns ndims = 1
[1,0]<stderr>:[SPLASH_LOG:0]  returns sizeRead = (32,1,1)
[1,3]<stderr>:[SPLASH_LOG:3]  returns ndims = 1
[1,3]<stderr>:[SPLASH_LOG:3]  returns sizeRead = (32,1,1)
[1,2]<stderr>:[SPLASH_LOG:2]  returns ndims = 1
[1,2]<stderr>:[SPLASH_LOG:2]  returns sizeRead = (32,1,1)
[1,1]<stderr>:[SPLASH_LOG:1]  returns ndims = 1
[1,1]<stderr>:[SPLASH_LOG:1]  returns sizeRead = (32,1,1)
[1,0]<stderr>:[SPLASH_LOG:0] closing parallel data collector
[1,1]<stderr>:[SPLASH_LOG:1] closing parallel data collector
[1,2]<stderr>:[SPLASH_LOG:2] readDataSet
[1,3]<stderr>:[SPLASH_LOG:3] readDataSet
[1,2]<stderr>:[SPLASH_LOG:2] DCDataSet::read (x)
[1,2]<stderr>:[SPLASH_LOG:2] 
[1,2]<stderr>: ndims         = 1
[1,2]<stderr>: logical_size  = (14435572,1,1)
[1,2]<stderr>: physical_size = (14435572,1,1)
[1,2]<stderr>: dstBuffer     = (662223,1,1)
[1,2]<stderr>: dstOffset     = (0,0,0)
[1,2]<stderr>: srcSize       = (662223,1,1)
[1,2]<stderr>: srcOffset     = (0,0,0)
[1,2]<stderr>:
[1,3]<stderr>:[SPLASH_LOG:3] DCDataSet::read (x)
[1,3]<stderr>:[SPLASH_LOG:3] 
[1,3]<stderr>: ndims         = 1
[1,3]<stderr>: logical_size  = (14435572,1,1)
[1,3]<stderr>: physical_size = (14435572,1,1)
[1,3]<stderr>: dstBuffer     = (685264,1,1)
[1,3]<stderr>: dstOffset     = (0,0,0)
[1,3]<stderr>: srcSize       = (685264,1,1)
[1,3]<stderr>: srcOffset     = (662223,0,0)
```

### PR DESCRIPTION
**Pull request by ax3l**
- fix #609 particle read during restart hangs if some GPUs have no particles and others do

### To Do
- [x] check at runtime

### CODE REVIEW

**Comment by ax3l** [comment]
@f-schmitt-zih can we write a minimal-libSplash test for that to see if the error comes from the HDF5 implementation?

**Comment by f-schmitt-zih** [comment]
Yes, I can do that

**Comment by ax3l** [comment]
also, I think @psychocoderHPC made a discovery by tapering the arguments for libSplash to read "one fake element" instead of NULL today and could restart the bunch example just before going home. he might can get an update to that tomorrow - so it really looks like it is something in libSplash.

**Comment by psychocoderHPC** [comment]
@ax3l, @f-schmitt-zih  I pushed the workaround to [this pull request](https://github.com/ax3l/picongpu/pull/3)
and opened a issue in [libSplash](https://github.com/ComputationalRadiationPhysics/libSplash/issues/148)

**Comment by f-schmitt-zih** [comment]
Should I still look into the repro case? I won't be able to do that before the weekend, though. However, I will fix the libsplash [issue](https://github.com/ComputationalRadiationPhysics/libSplash/issues/148).

**Comment by ax3l** [comment]
That's great, thank you!
I think it might be useful just to add the test to debug and fix the [libSplash issue](https://github.com/ComputationalRadiationPhysics/libSplash/issues/148) itself.

**Comment by ax3l** [comment]
@f-schmitt-zih can you pls comment/merge this pull if you find the time? :)

**Comment by f-schmitt-zih** [comment]
The change is fine!

**Comment by ax3l** [comment]
lol, "Closed with unmerged commits" @f-schmitt-zih ;)

**Comment by f-schmitt-zih** [comment]
I should really work less...

**Comment by ax3l** [comment]
~~work~~ party :santa:

**Comment by ax3l** [comment]
@f-schmitt-zih @psychocoderHPC tested :shipit:

**Comment by ax3l** [comment]
update: I just ran a 2 species example from `dev` where this patch still does not help: the restart does not hang but I loose around 75% to 95% of the particles per species.

File: src/picongpu/include/plugins/hdf5/restart/LoadParticleAttributesFromHDF5.hpp [hunk 1]
```diff
@@ -76,7 +76,20 @@ struct LoadParticleAttributesFromHDF5
...
+        else
+        {
+            /** This is a workaround to avoid the bug that processes hangs
+             *  if some call `DataCollector::read()` where elements to read is
+             *  zero and the destination pointer for the read data is set to `NULL`.
+             *  see: - `PIconGPU`  issu: https://github.com/ComputationalRadiationPhysics/picongpu/pull/611
+             *       - `libSplash` issu: https://github.com/ComputationalRadiationPhysics/libSplash/issues/148
+             * \todo: please remove this workaround after the libsplash bug is fixed
+             */
+            tmpArray = new ComponentType[1];
```
Comment (ax3l) [inline src/picongpu/include/plugins/hdf5/restart/LoadParticleAttributesFromHDF5.hpp]: **Update:** I just run a test with two species, where the sim did not hang but lost 75% and 95% of the particles during restart respectively. That work-around seems to be invalid, I'll test the new `1.2.4` libSplash patch.

Comment (ax3l) [inline src/picongpu/include/plugins/hdf5/restart/LoadParticleAttributesFromHDF5.hpp]: **Update:** no, the work-around is valid but we have an other bug with static load balancing restarts right now #637

**Pull request merged**